### PR TITLE
Feat(Hashicorp Vault): Adds Hashicorp Vault as a crypto key store

### PR DESCRIPTION
**Pull request by pdt256**

### CODE REVIEW

File: pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore.go [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+		log.Printf("%d: %s", response.StatusCode, body)
+
+		return fmt.Errorf("unable to save")
+	}
+
+	return nil
+}
+
+func (h *hashicorpVaultKeyStore) Delete(subjectID string) error {
+	jsonString := `{"versions":[1,2]}`
```
Comment (tomParty) [inline pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore.go:108]: env var with default?

File: pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore.go [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+
+func (h *hashicorpVaultKeyStore) getDestroyPath(subjectID string) string {
+	return fmt.Sprintf("%s/v1/secret/destroy/%s", h.config.Address, subjectID)
+}
+
+func (h *hashicorpVaultKeyStore) SetTimeout(duration time.Duration) {
+	h.httpClient.Timeout = duration
+}
+
+func (h *hashicorpVaultKeyStore) SetTransport(transport http.RoundTripper) {
```
Comment (tomParty) [inline pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore.go:167]: 🔥

File: pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+	"github.com/stretchr/testify/require"
+
+	"github.com/inklabs/rangedb/pkg/crypto"
+	"github.com/inklabs/rangedb/pkg/crypto/cryptotest"
+	"github.com/inklabs/rangedb/pkg/crypto/provider/vaultkeystore"
+)
+
+func TestHashicorpVault_VerifyEngineInterface(t *testing.T) {
+	config := getConfigFromEnvironment(t)
+	const iv = "1234567890123456"
```
Comment (tomParty) [inline pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go:21]: iv?

File: pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore.go [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+		return "", err
+	}
+	request.Header.Set("X-Vault-Token", h.config.Token)
+
+	response, err := h.httpClient.Do(request)
+	if err != nil {
+		return "", err
+	}
+
+	var resp saveResponse
```
[the remark this answers is not in the document]
Comment (pdt256) [reply pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore.go:49]: Get and Post both have the same response json schema. I'll rename this struct to `secretResponse`

File: pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go [hunk 5]
```diff
@@ -0,0 +1,231 @@
...
+					Body:       ioutil.NopCloser(strings.NewReader("{}")),
+				}, nil
+			}))
+
+			// When
+			err = vaultCrypto.Set(subjectID, encryptionKey)
+
+			// Then
+			require.EqualError(t, err, "unable to save")
+		})
```
Comment (tomParty) [inline pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go:153]: curious the error for when set if an already existed key?

File: pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore.go (hunk 1 above)
Comment (pdt256) [reply pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore.go:108]: I'll change to a package constant

File: pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go [hunk 6]
```diff
@@ -0,0 +1,231 @@
...
+				}, nil
+			}))
+
+			// When
+			err = vaultCrypto.Delete(subjectID)
+
+			// Then
+			require.EqualError(t, err, "unable to delete encryption key")
+		})
+	})
```
Comment (tomParty) [inline pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go:201]: curious if delete key when no key present, the response

File: pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go (hunk 3 above)
Comment (pdt256) [reply pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go:21]: initialization vector for AES encryption. I forgot to remove encryption from this keystore. Great catch @tomParty

File: pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go [hunk 7]
```diff
@@ -0,0 +1,231 @@
...
+
+func NewStubTransport(roundTrip func(request *http.Request) (*http.Response, error)) *stubTransport {
+	return &stubTransport{
+		roundTrip: roundTrip,
+	}
+}
+
+func (s *stubTransport) RoundTrip(request *http.Request) (*http.Response, error) {
+	return s.roundTrip(request)
+}
```
[the remark this answers is not in the document]
Comment (pdt256) [reply pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go:231]: 👍 

I still don't know how to cover this line:
```go
payload := dataPayload{Data: map[string]string{
	"key": key,
}}
jsonString, err := json.Marshal(payload)
if err != nil {
	return err
}
```

[the remark this answers is not in the document]
Comment (tomParty) [reply pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go:231]: thats tough, might want to do a empty string check on `key`

[the remark this answers is not in the document]
Comment (pdt256) [reply pkg/crypto/provider/vaultkeystore/hashicorp_vault_keystore_test.go:231]: fixed in: 34c86e367620f62bc02bed63dbcdd9b83ab9d62f

**Pull request merged**